lib/vmapi-client.js

```diff
@@ -9,11 +9,120 @@
  */
 
 /*
- * vmapi-client.js
+ * This module contains the VmapiClient object which has functions for
+ * interaction with VMAPI.
+ *
  */
 
-var restify = require('restify');
+var restify = require('restify-clients');
+
+
+/*
+ * VMAPI has default values for some fields when the fields are not actually
+ * in the VM objects in Moray. For some of those fields such as:
+ *
+ *  autoboot: null
+ *  brand: null
+ *  firewall_enabled: false
+ *  nics: []
+ *  platform_build_stamp: null
+ *  resolvers: []
+ *  server_uuid: null
+ *  snapshots: []
+ *  state: null
+ *  tags: {}
+ *  zone_state: null
+ *  zonepath: null
+ *
+ * vmadm *also* always has these, so any time they're different we still want to
+ * do an update. The ones VMAPI always includes that are *not* always in vmadm
+ * objects are put into VMAPI_UNSET_FIELDS here so that we can detect their
+ * default values and not force an update when the vmadm object does not have
+ * these but the VMAPI object does and has the default value.
+ *
+ * If we did do that update, we'd be updating such a VM on every restart since
+ * VMAPI would always give us an object with these properties even if the
+ * correct object were in Moray.
+ *
+ * We have also excluded:
+ *
+ *  customer_metadata: {}
+ *  internal_metadata: {}
+ *
+ * since these are not in the COMPARISON_FIELDS above. If they're added we'll
+ * need to do a bit more complicated comparison since we'll have to treat {}
+ * from VMAPI and not having either of these from vmadm as equivalent.
+ *
+ */
+ // TODO XXX: add a note in the VMAPI code that if the list is updated and more
+ // fields are added, we come back here and update this too.
+var VMAPI_UNSET_FIELDS = {
```

Comment:
I'm wondering if there's a way to test that this list is up to date with VMAPI's. Would a test that does the following:

1. Create a new VM
2. Start vm-agent and check that that VM was not updated by vm-agent (because it hasn't changed since its creation).
3. Restart vm-agent and perform the same check

test that behavior?